File: android/app/src/test/java/io/nexy/android/service/SpokenOutputTest.kt
package io.nexy.android.service

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Test

class SpokenOutputTest {
    @Test
    fun `sanitizer removes code commands urls and markdown`() {
        val input = """
            # Result
            Use **the safe path** from [the guide](https://example.com/guide).
            `inlineCode()`
            ```kotlin
            dangerous()
            ```
            npm run build
            Keep this sentence.
        """.trimIndent()

        val spoken = sanitizeForSpeech(input)

        assertEquals("Result Use the safe path from the guide. Keep this sentence.", spoken)
        assertFalse(spoken.contains("example.com"))
        assertFalse(spoken.contains("dangerous"))
        assertFalse(spoken.contains("npm"))
    }

    @Test
    fun `quick recap is deterministic bounded and sentence aware`() {
        val input = "First result is complete. Second detail is intentionally long. Third detail should not be included."

        assertEquals("First result is complete.", createQuickRecap(input, 40))
        assertEquals(createQuickRecap(input, 40), createQuickRecap(input, 40))
    }

    @Test
    fun `quick recap truncates a single long sentence`() {
        assertEquals("abcd…", createQuickRecap("abcdefghij", 5))
    }

    @Test
    fun `settings normalization matches desktop bounds`() {
        val normalized = normalizeSpokenOutputSettings(
            SpokenOutputSettings(voiceId = "", rate = 9f, pitch = Float.NaN),
        )

        assertEquals(null, normalized.voiceId)
        assertEquals(2f, normalized.rate)
        assertEquals(1f, normalized.pitch)
    }
}

File: src/renderer/__tests__/spoken-output.test.ts
import { describe, expect, it } from 'vitest'
import {
  createQuickRecap,
  DEFAULT_SPOKEN_OUTPUT_SETTINGS,
  readSpokenOutputSettings,
  sanitizeForSpeech,
  SPOKEN_OUTPUT_SETTINGS_KEY,
} from '../lib/spoken-output'

describe('spoken output', () => {
  it('removes code, commands, URLs, and Markdown syntax', () => {
    const input = [
      '## Result',
      'Use **the safe option** at [Nexy](https://nexy.test).',
      '```ts',
      'const secret = "do not read this"',
      '```',
      'npm run build',
      'Then open https://example.test/path.',
    ].join('\n')

    const result = sanitizeForSpeech(input)

    expect(result).toBe('Result Use the safe option at Nexy. Then open')
    expect(result).not.toContain('secret')
    expect(result).not.toContain('npm')
    expect(result).not.toContain('https')
    expect(result).not.toMatch(/[*#`]/)
  })

  it('creates a deterministic recap without a model request', () => {
    const input = 'First result is complete. Second result explains the details. Third result is extra.'
    expect(createQuickRecap(input, 55)).toBe('First result is complete.')
    expect(createQuickRecap(input, 55)).toBe(createQuickRecap(input, 55))
  })

  it('clamps persisted settings and defaults privacy-sensitive values', () => {
    const storage = {
      getItem: (key: string) => key === SPOKEN_OUTPUT_SETTINGS_KEY
        ? JSON.stringify({ rate: 9, pitch: 0.1, offlineOnly: false, autoPlay: true, voiceUri: 'voice-1' })
        : null,
    }
    expect(readSpokenOutputSettings(storage)).toEqual({
      engine: 'system',
      voiceUri: 'voice-1',
      supertonicSpeakerId: 0,
      supertonicLanguage: 'auto',
      rate: 2,
      pitch: 0.5,
      offlineOnly: false,
      autoPlay: true,
    })
  })

  it('uses safe defaults when persisted settings are malformed', () => {
    expect(readSpokenOutputSettings({ getItem: () => '{bad json' }))
      .toEqual(DEFAULT_SPOKEN_OUTPUT_SETTINGS)
  })
})
